tests/functional/test_wheel.py

```diff
@@ -210,6 +210,7 @@ def test_wheel_package_with_latin1_setup(script, data, common_wheels):
     assert 'Successfully built SetupPyUTF8' in result.stdout
 
 
+@pytest.mark.xfail(reason="PEP 518 does not support links")
```

Comment:
Wait... Why?

Reply:
More specifically, this version does not support dependencies that don't have binary (wheel) files available. The reason why is explained in the initial post.

Reply:
The current implementation potentially allows pip to spawn multiple sub-processes recursively to install build_dependencies.  Solving that problem is too complex until we have the resolver, and the practical benefits will be realized with this implementation (think numpy, which depends on Cython, which has wheels available).

Reply:
> More specifically, this version does not support dependencies that don't have binary (wheel) files available. 

OK... so this version isn't mergeable because it doesn't support scenarios pip currently supports? Am I understanding that right? Or are you saying "pip has a bug in situation X", in which case let's fix that bug first rather than proposing code which stops pip working at all in situation X. If necessary, submit an isolated PR that says "pip has a catastrophic bug in situation X - detect that situation and refuse to proceed". Then once that's in, this PR can simply ignore situation X, leaving it as a future extension.

This is another example where you're including too much in one PR. You need to make your PRs *much* more granular. And yes, that does mean that work on later stages has to wait till earlier stages are thrashed out and merged. But that's better than a huge, un-reviewable mess that never gets merged.

Reply:
> this version isn't mergeable because it doesn't support scenarios pip currently supports

> let's fix that bug first rather than proposing code which stops pip working at all in situation X.

How can I say this? Situation X is rare. In fact, I know of no actual scenarios where situation X would actually occur. In addition, situation X is only supported because at the time, people did not consider the full implications of supporting it. Correctly supporting situation X requires either:

1. An entirely new refactoring and/or rewrite of key areas (correct).
2. Recursive process spawning (unmonitored and uncontrolled), which is what the current implementation does (this is horrible).

In contrast, there is another situation: situation Y. Situation Y, in contrast to situation X, is widely used and important. 

This PR drops the recursive process spawning (very bad), removing support for situation X (which would actually *maybe* be used once) and adds support for situation Y (widely used).

> But that's better than a huge, un-reviewable mess that never gets merged.

Now that would be implementing situation X correctly; something that even the original implementer avoided.

Reply:
More specifically, I can rattle off a list of at least three major projects (scipy, scikit-learn, numpy) that would use situation Y and don't care about situation X and quite a few more minor projects. I can think of none that would need situation X.

Reply:
Apologies by the way for the abstract "X" and "Y". I didn't really follow the actual problem, so I simplified (or not...)

As I say, if X is rare, then a PR that identifies it and explicitly reports a "we don't handle this" error should be fairly uncontroversial. So submit that, get it accepted, and then this PR can ignore that case in safety. By merging the two steps into one PR, we end up muddling the issues. Also, a PR that says "drop support for X" would *have* to clearly explain what X was, and I'd be able to avoid all these clumsy circumlocutions :-)

Reply:
> So submit that, get it accepted, and then this PR can ignore that case in safety.

Will do. Communication is good.